packages/ds-sam-sdk/src/auction.ts

```diff
@@ -0,0 +1,181 @@
+import { AuctionData, AuctionResult, AuctionValidator } from './types'
+import { AuctionConstraints } from './constraints'
+import { Debug } from './debug'
+
+const logValidators = (validators: AuctionValidator[]) => {
+  console.log('validators -----------------------------')
+  for (const validator of validators) {
+    console.log(validator.voteAccount, validator.revShare.totalPmpe, validator.auctionStake.marinadeMndeTargetSol, validator.auctionStake.marinadeSamTargetSol)
+  }
+  console.log('----------------------------- validators')
+}
+
+const EPSILON = 1e-4
+
+export class Auction {
+
+  constructor (private data: AuctionData, private constraints: AuctionConstraints, private debug: Debug) { }
+
+  distributeMndeStake () {
+    this.constraints.updateStateForMnde(this.data)
+    this.debug.getVoteAccounts().forEach((voteAccount) => {
+      const constraints = this.constraints.getValidatorConstraints(voteAccount)
+      this.debug.pushValidatorEvent(voteAccount, `MNDE start constraints: ${constraints ? `${JSON.stringify(constraints.map(constraint => ({ ...constraint, validators: constraint.validators.length})))}` : 'NULL'}`)
+    })
+
+    let remEligibleValidators = this.data.validators.filter(validator => validator.mndeEligible)
+
+    while (remEligibleValidators.length > 0) {
+      const eligibleValidators = new Map(remEligibleValidators.map(validator => [validator.voteAccount, validator]))
+      const eligibleVoteAccounts = new Set(eligibleValidators.keys())
+      const { cap: evenDistributionCap } = this.constraints.getMinCapForEvenDistribution(eligibleVoteAccounts)
+      console.log("MNDE distributing", evenDistributionCap, "to every validator in the group", eligibleValidators.size)
+
+      for (const validator of eligibleValidators.values()) {
+        validator.auctionStake.marinadeMndeTargetSol += evenDistributionCap
+        this.data.stakeAmounts.marinadeRemainingMndeSol -= evenDistributionCap
+      }
+
+      // logValidators(remEligibleValidators)
+
+      this.constraints.updateStateForMnde(this.data)
+      remEligibleValidators = remEligibleValidators.filter((validator) => {
+        const validatorCap = this.constraints.findCapForValidator(validator)
+        if (validatorCap < EPSILON) {
+          console.log('MNDE removing validator', validator.voteAccount, 'from the group because the cap has been reached')
+          return false
+        }
+        return true
+      })
+
+      if (this.data.stakeAmounts.marinadeRemainingMndeSol < EPSILON) {
+        console.log("MNDE No stake remaining to distribute")
+        break
+      } else {
+        console.log("MNDE Stake remaining", this.data.stakeAmounts.marinadeRemainingMndeSol)
+      }
+    }
+  }
+
+  distributeSamStake () {
+    this.data.validators.sort((a, b) => b.revShare.totalPmpe - a.revShare.totalPmpe)
```

Comment:
Correct, this should no longer be needed due to how `findNextPmpeGroup` works.

I think earlier when discussing the auction we said it would make sense to have the outputs ordered in _some_ way, so the results would be somehow "stable" string-wise. Not sure if PMPE is the best choice for this specific goal (there are still validators in the same PMPE group so we're basically relying on how stable the sort is), sorting alphabetically would be clearer, but having the highest PMPE validators at the top of the result probably makes some logical sense.

If you'd prefer to change it to an alphabetical (or other) sort or remove the sorting completely, please let me know. Otherwise, I'd probably keep it this way for now.

Reply:
Let's keep as is, then.

If we ever want to do diffs, we can always sort in different way just-in-time